### ISSUE TITLE
[5] Distribute transactions over full 3 seconds to UI

### ISSUE THREAD

**Issue by wmbutler** [bitshares/bitshares-ui#1985]
**Is your feature request related to a problem? Please describe.**
This is a more of a perception problem, but since perception is an important factor when talking about markets, I think it's worth discussion. @startailcoon and @ryanRfox discussed this with me while on a call and I'd like to formalize it here.

BTS blocks are 3 seconds but there can be many transactions in a single block. Our wallet rather simplistically updates the Exchange every 3 seconds, but in the eyes of a casual user it makes the exchange feel very slow. I am proposing that for each block we:

- count the number of transactions and divide 3 by that number of transactions to determine the update interval. We then deliver 1 transaction to the frontend per interval.

Example:

30 transactions are contained in a block
3/30 = .1

We deliver 1 transaction to the UI every .1 seconds. This will create a uniform flow to the exchange and also to user assets, portfolio value, etc. This coupled with our home page featuring active markets might have the intended effect of showing a constant stream of liquidity.

**Comment by sschiessl-bcp** [comment]
When the exchange is open, incoming orders are pushed from the backend. Possible that this might need internal caching.

Either way, this option should be configurable. I for example would prefer to be up-to-date rather than get a perceived fake distribution of orders.

**Comment by svk31** [comment]
Delaying orderbook updates by that much is a bad idea since it keeps information from users, so I'm strongly opposed to doing this. Also, the orderbook is not simplistically updated every 3s, it's updated whenever a notification is dispatched by the API.

In fact there is already a queuing system, but it was implemented for performance reasons. It batches updates and dispatches them at most every 0.5s. I would not want to take that higher than 0.5s however.

**Comment by sschiessl-bcp** [comment]
> In fact there is already a queuing system, but it was implemented for performance reasons. It batches updates and dispatches them at most every 0.5s. I would not want to take that higher than 0.5s however.

I never noticed. Could you point me to it? Just curious now

**Comment by svk31** [comment]
https://github.com/bitshares/bitshares-ui/blob/develop/app/actions/MarketsActions.js#L161

**Comment by wmbutler** [comment]
> In fact there is already a queuing system, but it was implemented for performance reasons. It batches updates and dispatches them at most every 0.5s. I would not want to take that higher than 0.5s however.

I'm suggesting we dispatch them at lower values. Not suggesting a higher value. This won't affect trades because the order matching happens on-chain right? I'm not sure I understand your concern but would like to better understand it.

I think the market platform will always suffer from the impression that it feels slow even in .5 second increments. Modern trading platforms stream hundreds of visually discernible transactions per second and I believe and end-users picks up on this.

**Comment by clockworkgr** [comment]
Fwiw , i think it's still impressive as it is now : https://wallet.bitshares.org/#/explorer/blocks

**Comment by wmbutler** [comment]
@clockworkgr @ryanRfox we could employ this strategy only on the market page if there is pushback. The other pages probably don't matter as much.

**Comment by sschiessl-bcp** [comment]
> Wondering if anyone has capacity to get this into a branch by 14 NOV? Yes, I understand it is short notice. BitShares have a booth at the Decentralized conference 15-16 NOV where me, @clockworkgr and other representatives hope to showcase the platform live. Really impressive to see those ~20+ tps stream on screen.

Alex is working on a custom dashboard for decentralized

**Comment by clockworkgr** [comment]
http://clients.clockwork.gr/decentralized/

This is for display only though...not sure how well and how it would work if feeding ALL components of the wallet

**Comment by clockworkgr** [comment]
@startailcoon  on mine, just click anywhere on the table...clcik again to unpause :P

**Comment by dot5enko** [comment]
@clockworkgr wouldn't this cause performance drop (on market page) as react components would redraw so frequently?

**Comment by clockworkgr** [comment]
no idea to be honest...possibly...

fwiw the one in mine is written in Vue...so it's not completely barebones either....we can always try it out

**Comment by sschiessl-bcp** [comment]
> @clockworkgr wouldn't this cause performance drop (on market page) as react components would redraw so frequently?

Good point. If `componentShouldUpdate` is used properly throughout I imagine no performance drop, so I would put that as requirement.

**Comment by clockworkgr** [comment]
i only know how to do it properly in Vue...can't help...

**Comment by wmbutler** [comment]
> @clockworkgr wouldn't this cause performance drop (on market page) as react components would redraw so frequently?

I doubt it. There will only be a few hundred per second max.

**Comment by startailcoon** [comment]
To better showcase the functionality of this I would suggest that we rework the "Explorer -> Blockchain" page with something like what is done on clients.clockwork.gr/decentralized/#/

If we redesign the `Info` and `Recent blocks` to look like what clockwork has done.

Right now it's not showing anything relevant since the blocks have so many transactions in them that they can't be visible.

**Comment by sschiessl-bcp** [comment]
I'd be down testing this feature, also from the architectural implementation on the Explorer. If it turns out feasible, can be extended to Exchange

**Comment by clockworkgr** [comment]
fwiw, the way it roughly works on my page is as follows....

Display a table component: 

Table rows are bound to an array of operations (each table row ->  1 operation object)

Start with an empty array.

On each incoming block, extract list of operations, add them in a different array as pending.

Run a setInterval function that pops one op from the pending array and pushes it to the displayed operations array with an interval of 3000/number_of_ops_in_block.
Truncate displayed operations array if more than 100 items

**Comment by wmbutler** [comment]
@clockworkgr want to implement this on the BlockChain Explorer page?

**Comment by clockworkgr** [comment]
> @clockworkgr want to implement this on the BlockChain Explorer page?

I'll have a look to see if I can...remember I'm not that well versed in React and my solution uses ES rather than the standard chain data

**Comment by startailcoon** [comment]
@clockworkgr did you have any good results on this on the recent blocks page?
